src/graphql/dataSource.js

```diff
@@ -0,0 +1,169 @@
+const {DataSource} = require('apollo-datasource');
+const {search} = require('../operations/search/search');
+const {getRequestInfo} = require('./requestInfoHelper');
+const {searchById} = require('../operations/searchById/searchById');
+const {logWarn} = require('../operations/common/logging');
+const async = require('async');
+
+// const DataLoader = require('dataloader');
+
+class FhirDataSource extends DataSource {
+    constructor() {
+        super();
+        // this.dataLoader = new DataLoader(keys => myBatchGetUsers(keys));
+        this.meta = [];
+    }
+
+    initialize(config) {
+
+        return super.initialize(config);
+    }
+
+    /**
+     * This functions takes a FHIR Bundle and returns the resources in it
+     * @param {{entry:{resource: Resource}[]}} bundle
+     * @return {Resource[]}
+     */
+    unBundle(bundle) {
+        if (bundle.meta) {
+            this.meta.push(bundle.meta);
+        }
+        return bundle.entry.map(e => e.resource);
+    }
+
+    /**
+     * This is to handle unions in GraphQL
+     * @param obj
+     * @param context
+     * @param info
+     * @return {null|string}
+     */
+    // noinspection JSUnusedLocalSymbols
+    // eslint-disable-next-line no-unused-vars
+    resolveType(obj, context, info) {
+        if (Array.isArray(obj)) {
+            if (obj.length > 0) {
+                // apollo does not seem to allow returning lists
+                // return obj.map(o => resolveType(o, context, info));
+                return obj[0].resourceType;
+            } else {
+                return null;
+            }
+        } else {
+            return obj.resourceType;
+        }
+    }
+
+    /**
+     * Finds a single resource by reference
+     * @param {Resource} parent
+     * @param args
+     * @param context
+     * @param info
+     * @param {{reference: string}} reference
+     * @return {Promise<null|Resource>}
+     */
+    async findResourceByReference(parent, args, context, info, reference) {
+        if (!(reference)) {
+            return null;
+        }
+        /**
+         * @type {string}
+         */
+        const typeOfReference = reference.reference.split('/')[0];
```

Comment:
Should we check if the split results in an array > 1 to prevent hitting a index out of range exception